csharp/src/Ice/OutputStream.cs

```diff
@@ -169,232 +111,171 @@ public byte[] Finished()
-        /// </summary>
-        /// <param name="encoding">The encoding version of the encapsulation.</param>
-        public void WriteEmptyEncapsulation(EncodingVersion encoding)
-        {
-            Protocol.checkSupportedEncoding(encoding);
-            WriteInt(6); // Size
-            WriteByte(encoding.major);
-            WriteByte(encoding.minor);
-        }
+            // Write the tagged member end marker if some tagged members were encoded. Note that the optional members
+            // are encoded before the indirection table and are included in the slice size.
+            if ((_current.SliceFlags & Protocol.FLAG_HAS_OPTIONAL_MEMBERS) != 0)
+            {
+                WriteByte(Protocol.OPTIONAL_END_MARKER);
+            }
 
-        /// <summary>
-        /// Writes a pre-encoded encapsulation.
-        /// </summary>
-        /// <param name="v">The encapsulation data.</param>
-        public void WriteEncapsulation(byte[] v)
-        {
-            if (v.Length < 6)
+            // Write the slice size if necessary.
+            if ((_current.SliceFlags & Protocol.FLAG_HAS_SLICE_SIZE) != 0)
             {
-                throw new EncapsulationException();
+                int sz = Pos - _current.SliceFirstMemberPos + 4;
+                RewriteInt(sz, _current.SliceFirstMemberPos - 4);
             }
-            expand(v.Length);
-            _buf.b.put(v);
-        }
 
-        /// <summary>
-        /// Determines the current encoding version.
-        /// </summary>
-        /// <returns>The encoding version.</returns>
-        public EncodingVersion GetEncoding()
-        {
-            return _encapsStack != null ? _encapsStack.encoding : _encoding;
-        }
+            if (_current.IndirectionTable?.Count > 0)
+            {
+                Debug.Assert(_format == FormatType.SlicedFormat);
+                _current.SliceFlags |= Protocol.FLAG_HAS_INDIRECTION_TABLE;
 
-        /// <summary>
-        /// Marks the start of a new slice for a class instance or user exception.
-        /// </summary>
-        /// <param name="typeId">The Slice type ID corresponding to this slice.</param>
-        /// <param name="compactId">The Slice compact type ID corresponding to this slice or -1 if no compact ID
-        /// is defined for the type ID.</param>
-        /// <param name="last">True if this is the last slice, false otherwise.</param>
-        public void StartSlice(string typeId, int compactId, bool last)
-        {
-            Debug.Assert(_encapsStack != null && _encapsStack.encoder != null);
-            _encapsStack.encoder.StartSlice(typeId, compactId, last);
-        }
+                WriteSize(_current.IndirectionTable.Count);
+                foreach (var v in _current.IndirectionTable)
+                {
+                    WriteInstance(v);
+                }
+                _current.IndirectionTable.Clear();
+                _current.IndirectionMap?.Clear();
```

Comment:
Why `IndirectMap?` here? AFAICT, it should always be non-null if IndirectTable is not null.

Reply:
You need to do that because it is declared as nullable `?`, you can avoid this here by adding an assert `Debug.Assert(_current.IndirectionMap != null)` 

The compiler will not complain after the assert

Reply:
IndirectionTable is also declared as a nullable, I guess the compiler is smart enough to not issue a warning because of the `if(_current.IndirectionTable?.Count > 0)`? Here, my point was that I didn't see when IndirectionMap could be null while IndirectionTable not being null... that said I found the reason in the write sliced data code (see bellow).

In general, if we know that the reference isn't null, shouldn't we use the null-forgiving operator instead of asserts?

Reply:
I think we should prefer the assert if the reference being null indicates a logic error, the assert will only run on debug so has no overhead for release builds. I think ! is similar to the assert in this case.

Null forgiving ? should be used when being null is an expected condition

Reply:
Note that the [null-forgiving operator](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-forgiving) is actually the ! operator while  ?. is the null-conditional operator. The null-forgiving operator has actually not runtime effects so asserts are still useful for null checks.

Reply:
`_IndirectionMap` can be null, when we write a UnknownSlicedClass. See WriteSlicedData. I'll add a comment.

Reply:
Both `Debug.Assert(_current != null)` and _current!.Xxx tell the compiler "I know _current is not null here", so stop issuing warnings about _current being possibly null. I find the Debug.Assert clearer.